File: Angular-App/adminportalapp/src/app/service/user.service.ts
import { Injectable } from "@angular/core";
import { environment } from "../../environments/environment";
import {
  HttpClient,
  HttpResponse,
  HttpErrorResponse,
  HttpEvent
} from "@angular/common/http";

import { from, Observable } from "rxjs";
import { User } from "../model/user";
import { JwtHelperService } from "@auth0/angular-jwt";
import { CustomHttpResponse } from '../model/custom-http-response';

@Injectable({
  providedIn: "root"
})
export class UserService {
  private host = environment.apiUrl;

  constructor(private http: HttpClient) {}

  public getUsers(): Observable<User[]> {
    return this.http.get<User[]>(`${this.host}/user/list`);
  }

  public addUser(formData: FormData): Observable<User> {
    return this.http.post<User>(`${this.host}/user/add`, formData);
  }

  public updateUser(formData: FormData): Observable<User> {
    return this.http.post<User>(`${this.host}/user/update`, formData);
  }

  public resetPassword(email: string): Observable<CustomHttpResponse> {
    return this.http.get<CustomHttpResponse>(`${this.host}/user/resetPassword/${email}`);
  }

  public updateProfileImage(
    formData: FormData
  ): Observable<HttpEvent<User>> {
    return this.http.post<User>(
      `${this.host}/user/updateProfileImage`,
      formData,
      { reportProgress: true, observe: "events" }
    );
  }

  public deleteUser(userId: number): Observable<CustomHttpResponse> {
    return this.http.delete<CustomHttpResponse>(`${this.host}/user/delete/${userId}`);
  }

  public addUsersToLocalCache(users: User[]): void {
    localStorage.setItem("users", JSON.stringify(users));
  }

  public getUsersFromLocalCache(): User[] {
    if(localStorage.getItem("users")) {
      return JSON.parse(localStorage.getItem('users'));
    } 
    return null;
  }

  public createUserFormData(loggedInUserName: string, user: User, profileImage: File): FormData {
    const formData = new FormData();
    formData.append('currentUsername', loggedInUserName);
    formData.append('firstName', user.firstName);
    formData.append('lastName', user.lastName);
    formData.append('username', user.username);
    formData.append('email', user.email);
    formData.append('role', user.role);
    formData.append('profileImage', user.profileImageUrl);
    formData.append('isActive', JSON.stringify(user.active));
    formData.append('isNonLocked', JSON.stringify(user.notLocked));
    return formData;
  }

}

File: admin-portal-service/src/main/java/com/example/microservices/adminportal/service/UserService.java
package com.example.microservices.adminportal.service;

import java.io.IOException;
import java.util.List;

import org.springframework.web.multipart.MultipartFile;

import com.example.microservices.adminportal.exception.model.EmailExistException;
import com.example.microservices.adminportal.exception.model.NotAnImageException;
import com.example.microservices.adminportal.exception.model.UserNotFoundException;
import com.example.microservices.adminportal.exception.model.UsernameExistException;
import com.example.microservices.adminportal.model.User;

public interface UserService {

	public User register(String firstName, String lastName, String username, String email)
			throws UsernameExistException, EmailExistException;

	public List<User> getUsers();

	User findUserByUsername(String username);

	User findUserByEmail(String email);

	User addNewUser(String firstName, String lastName, String username, String email, String role, boolean isNonLocked,
			boolean isActive, MultipartFile profileImage) 
			throws UsernameExistException, EmailExistException, IOException, NotAnImageException;

	User updateUser(String currentUsername, String newFirstName, String newLastName, String newUsername,
			String newEmail, String role, boolean isNonLocked, boolean isActive, MultipartFile profileImage)
			throws UsernameExistException, EmailExistException, IOException, NotAnImageException;

	void deleteUser(String username);

	void resetPassword(String email) throws UserNotFoundException;

	User updateProfileImage(String username, MultipartFile profileImage) 
			throws UsernameExistException, EmailExistException, IOException, NotAnImageException;

}
